modules/core/src/matrix_transform.cpp

```diff
@@ -812,6 +812,71 @@ void flip( InputArray _src, OutputArray _dst, int flip_mode )
         flipHoriz( dst.ptr(), dst.step, dst.ptr(), dst.step, dst.size(), esz );
 }
 
+// TODO: improve it with in-place operation,
+//  see discussion https://github.com/opencv/opencv/pull/22898#discussion_r1042537867
+static void
+flipNDImpl(const uchar* src, const int* shape, const size_t* step, uchar* dst, int axis)
+{
+    int total = 1;
+    for (int i = 0; i < axis; ++i)
+        total *= shape[i];
+
+    int shape_at_axis = shape[axis];
+    size_t step_at_axis = step[axis];
+    for (int i = 0; i < total; ++i)
+    {
+        // retrieve offset for flip location
+        int t = i;
+        int offset = 0;
+        for (int j = axis - 1; j >= 0; --j)
+        {
+            int idx = t / shape[j];
+            offset += int(t - idx * shape[j]) * step[j];
+            t = idx;
+        }
+
+        for (int j = 0, k = shape_at_axis - 1; j < shape_at_axis; ++j, --k)
+            std::memcpy(dst + offset + j * step_at_axis, src + offset + k * step_at_axis, step_at_axis);
+    }
+}
+
+void flipND(InputArray _src, OutputArray _dst, int _axis)
+{
+    CV_INSTRUMENT_REGION();
+
+    // support Mat only
+    CV_Assert_N(_dst.isMat(), _src.isMat());
+
+    Mat src = _src.getMat();
+    int ndim = src.dims;
+    const auto shape = src.size.p;
+    const auto step = src.step.p;
+
+    // verify axis
+    int axis = (_axis + ndim) % ndim;
+    CV_CheckLT(axis, ndim, "flipND: unexpected axis >= ndim");
```

Comment:
Which values of `axis` do we want to support?
Check looks very weak (invalid usage is just ignored in "some way").